Adds usage of memory-mapping where available to read files; fixes some handle leaks; improves C++98-compatibility

NBickford-NV opened the pull request:
Hi Syoyo! Thanks for making TinyEXR.

During static analysis, we found a couple of places where file handles could be leaked (i.e. `fclose()` not called in some situations). While fixing these, I noticed that there were TODO items about adding memory-mapping (referenced in [PR #182](https://github.com/syoyo/tinyexr/pull/182#issuecomment-1312829463)). In this pull request, I've tried to combine the solutions to these two topics:

* Adds memory-mapping code for reading files
  * I've based this code off [viewer.cc](https://github.com/tinyobjloader/tinyobjloader/blob/4a02d274ae1836b40adaa603fbafa97a76c2abc7/experimental/viewer.cc#L90), with additional checks around error handling.
  * It includes code paths for Windows (it should work as early as Windows XP) and POSIX APIs. If neither is available, it falls back to C file I/O calls with additional error-checking (e.g. checking that `ftell()` succeeds - it's difficult, but possible, to get it to return a negative number!)
  * The code to memory-map a file is somewhat long, but moving the file I/O calls to a common place make TinyEXR shorter overall as a result.
* Fixes some file handle leaks
  * For instance, if the file was shorter than `kEXRVersionSize`, calling `ParseEXRVersionFromFile()` would fail to call `fclose()` at https://github.com/syoyo/tinyexr/blob/2a3ffbfab2112d541aa464fd749c528533bef172/tinyexr.h#L8499-8500.
  * I've done this generally by moving file opening to a `MemoryMappedFile` object; constructing a `MemoryMappedFile` opens a file for memory mapping, and closes the handle whenever it is destroyed. This ensures file handles are always closed on all code paths. When saving, I've added `fclose()` calls to the other code paths.
* Zero-initializes `ChannelInfo` in `ReadChannelInfo` - I noticed this left `requested_pixel_type` uninitialized.

I've run the unit tests on Windows under Visual Studio 2017 and 2022, and on Ubuntu Linux on WSL2 using clang 12 and the latest version of the openexr-images repository. All 36 test cases pass on Windows; on Linux, I see some errors from UndefinedBehaviorSanitizer, though they don't seem to be due to these changes.

I've only modified the code to read files here; if it would be good for `MemoryMappedFile` to also be used in the save paths, let me know and I'd be happy to modify it to support those! I could also modify `ParseEXRVersionFromFile` - with `MemoryMappedFile`, it becomes

```c++
int ParseEXRVersionFromFile(EXRVersion *version, const char *filename) {
  if (filename == NULL) {
    return TINYEXR_ERROR_INVALID_ARGUMENT;
  }

  MemoryMappedFile file(filename);
  if (!file.valid()) {
    return TINYEXR_ERROR_CANT_OPEN_FILE;
  }

  if (file.size < tinyexr::kEXRVersionSize) {
    return TINYEXR_ERROR_INVALID_FILE;
  }

  return ParseEXRVersionFromMemory(version, file.data,
                                   tinyexr::kEXRVersionSize);
}
```

but on platforms where neither Windows nor POSIX APIs are available, it would read the full file when it only needs to read the first 8 bytes.

Thanks again!

NBickford-NV committed:
Use memory-mapping where available to read files. Fixes leaked file handles in some error cases. Zero-initializes ChannelInfo to avoid a potential issue where ReadChannelInfo() leaves ChannelInfo::requested_pixel_type uninitialized.

Signed-off-by: Neil Bickford <nbickford@nvidia.com>

NBickford-NV committed:
Use TINYEXR_USE_WIN32_MMAP and TINYEXR_USE_POSIX_MMAP instead of _WIN32 when declaring fields in MemoryMappedFile

NBickford-NV committed:
Fix integer overflow when running test_tinyexr on starmap_2020_64k.exr

NBickford-NV commented:
I've added one more commit to this - commit e5c45f2 adds a fix for an issue where the width and height of the image were being multiplied before casting to a size_t; when running test_tinyexr on the 32768 x 65536 starmap_2020_64k from PR #182, this resulted in trying to allocate a vector with 2^64 - 2^31 elements. With this fix, `test_tinyexr starmap_2020_64k.exr` now succeeds on a computer with enough RAM!

syoyo commented:
@NeilBickford-NV Super Awesome! Let me give some time to review PR

Builtin-mmap feature is nice. I think we are better to provide new API for it, since current API design is too old and `LoadEXRImageFromFile` API does not have an ability to pass extra parsing options

It'd be something like...

```
auto mmap_file = EXRMemoryMappedFile(...); // Add `EXR` prefix for C-API
int ret = LoadEXRFromMemoryMappedFile(mmap_file, ...);
```

and add new error code if required.

NBickford-NV committed:
Match type of ftell() to avoid casting -1 to a size_t; free handle if file is too short; correctly zero-initialize ChannelInfo.

NBickford-NV commented:
Quick update: I added some quick fixes to this pull request in commit a282306:
* In `ParseEXRVersionFromFile()`, checks that `ftell()` doesn't return a negative number (which occurs when it encounters an error), and releases the file handle if the file is too short.
* Fixes the way I was zero-initializing `ChannelInfo`.
Thanks again!

NBickford-NV committed:
Add checks to fseek() calls for error codes

syoyo commented:
👍

NBickford-NV committed:
Remove computation of file size in ParseEXRVersionFromFile(); because this function only needs to know if reading kEXRVersionSize bytes worked, we can avoid a call to fseek(fp, 0, SEEK_END).

syoyo commented:
@NeilBickford-NV Confirmed `starmap_2020_64k.exr` can read correctly(64GB consumed out of 128 GB CPU mem PC).

BTW there are some clang warnings when we raise the warning level(on Ubuntu + clang-12 )

https://github.com/syoyo/tinyexr/blob/3c8e0cb7b406adc1bf9abe31510b2d72849ca6d2/Makefile#L2

We can add `-Wno-c++98-compat` to suppress initializer list warning, but at least it'd be better to add explicit casting in C++ code to suppress `-Wsign-conversion`.

```
$ clang++ -fsanitize=address -Wno-padded -Weverything -Werror -Wall -Wextra -std=c++11 -g -O0 -DTINYEXR_USE_PIZ=1 -I./deps/miniz -o test_tinyexr test_tinyexr.cc miniz.o
In file included from test_tinyexr.cc:20:
./tinyexr.h:1191:25: error: constructor call from initializer list is incompatible with C++98 [-Werror,-Wc++98-compat]
    ChannelInfo info = {};
                        ^
./tinyexr.h:6118:31: error: implicit conversion changes signedness: 'int' to 'unsigned long' [-Werror,-Wsign-conversion]
                              i;
                              ^
./tinyexr.h:6116:51: error: implicit conversion changes signedness: 'int' to 'unsigned long' [-Werror,-Wsign-conversion]
            const size_t ii = exr_image.tiles[it].offset_x *
                              ~~~~~~~~~~~~~~~~~~~~^~~~~~~~ ~

...
```

NBickford-NV committed:
Fixes warnings and errors when compiling for C++98:
- Fixes error when processing `#if TINYEXR_HAS_CXX11` when compiling with C++98
- Zero-initializes ChannelInfo using `memset` instead of ` = {}`
- Avoids implicit signed-to-unsigned casts
- Moves default initializers for MemoryMappedFile to its constructor
- Replaces `nullptr` with `NULL`
- Moves deletion of move and copy constructors behind a CXX11 flag
- Uses `std::map::insert` instead of `std::map::emplace` when C++11 is not available

NBickford-NV commented:
Thank you @syoyo ! I've added commit https://github.com/syoyo/tinyexr/pull/185/commits/606011e9df3f2f57bdbb5eb2f08df3b6e2203a84, which fixes all errors when compiling with clang++ 12 and

```
CXX = clang++
CXXFLAGS ?= -fsanitize=address -Wno-padded -Weverything -Werror -Wall -Wextra -std=c++11 -g -O0 -DTINYEXR_USE_PIZ=1
```

and

```
CXXFLAGS ?= -fsanitize=address -Wno-padded -Weverything -Werror -Wall -Wextra -std=c++98 -g -O0 -DTINYEXR_USE_PIZ=1
```

As a result, tinyexr should now be C++98-compatible - there were a few code changes outside of the memory-mapping code, but I went ahead and fixed the errors there as well.

Edit: Ah - `TChannelInfo` has an `std::string` member, so `memset(&info, 0, sizeof(info))` isn't valid! I'll have a fix for this shortly.

NBickford-NV committed:
Correct zero-initialization of ChannelInfo; fix errors on GCC

NBickford-NV committed:
Fix compilation error on C++11 gcc (due to clang-specific warning)

NBickford-NV commented:
I believe this pull request should now work with GCC and clang in `-std=c++98` and `-std=c++11` modes, as well as Visual Studio (in C++14 mode) - thanks again!

syoyo commented:
@NeilBickford-NV Super awesome! Thanks! Merged!

MemoryMappedFile feature seems working well in current TinyEXR API , so I accept your PR without any modification.

TODO: Consider providing separate API using mmap feature as commended in https://github.com/syoyo/tinyexr/pull/185#issuecomment-1359109355  for people who want to use mmap feature explicitly.

NBickford-NV commented:
Thank you @syoyo !
